GCSProxyServer.go

```diff
@@ -0,0 +1,256 @@
+package main
+
+import (
+	"awesomeProject/utils"
+	"awesomeProject/writers"
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+)
+
+type UploadSession struct {
+	uploadUrl         string
+	gcsClient         *utils.GcsClient
+	resumeOffset      int64
+	isAborted         bool
+	isCompleted       bool
+	activeConnections int
+	bucketName        string
+	objectName        string
+	sessionCtx        context.Context
+	cancelFunc        context.CancelFunc
+}
+
+func startGCSProxyServer(ctx context.Context, listenAddress string) error {
+	var (
+		uploadSessions      = make(map[string]*UploadSession)
+		uploadSessionsMutex sync.Mutex
+	)
```

Comment:
These are global variables in disguise + you need to pass them around to many functions.
```
type GcsProxyServer struct {
    uploadSessions map[string]*UploadSession
    lock           sync.Mutex
}

func (srv *GcsProxyServer) Serve(ctx context.Context, l net.Listener) {...}

func (srv *GcsProxyServer) handleConnection(ctx context.Context, conn net.Conn) {...}
```

Also, avoid overly long names: `lock` is as good as `uploadSessionsMutex`, but so much easier to read.